### PR TITLE
Ignore bad records in clients engine

### PR DESCRIPTION
**Pull request by JN3141**
Makes the clients engine check for tombstones and any deserialisation errors when receiving a client record, and ignores them if either are present

This PR addresses issue #1801

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
  - Added a test that covers two scenarios: one where there is a tombstone, and another where the JSON only has garbage fields that do not fit the schema
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/sync15/src/clients/engine.rs [hunk 1]
```diff
@@ -73,7 +73,16 @@ impl<'a> Driver<'a> {
             // clients collection, so we don't check for `is_tombstone`.
             // https://github.com/mozilla/application-services/issues/1801
             // tracks deleting these from the server.
-            let client: ClientRecord = payload.into_record()?;
+            if payload.is_tombstone() {
+                continue;
+            }
+
+            let client_res: Result<ClientRecord, serde_json::Error> = payload.into_record();
```
Comment (mhammond) [inline components/sync15/src/clients/engine.rs:80]: You can probably just use a `match` here - something like:
```
let client = match payload.into_record() {
    Some(client) => client,
    Err(_) => continue,
```

(although logging that we did skip probably makes sense)

File: components/sync15/src/clients/engine.rs [hunk 2]
```diff
@@ -73,7 +73,16 @@ impl<'a> Driver<'a> {
             // clients collection, so we don't check for `is_tombstone`.
             // https://github.com/mozilla/application-services/issues/1801
             // tracks deleting these from the server.
-            let client: ClientRecord = payload.into_record()?;
+            if payload.is_tombstone() {
+                continue;
```
Comment (mhammond) [inline components/sync15/src/clients/engine.rs:77]: it probably makes sense to log that we ignored a payload here?

**Review by mhammond — CHANGES_REQUESTED** [review]
Thanks for jumping in here! This looks great, although I think we can clean up the error handling, and unfortunately the CHANGES note went stale already.

**Comment by JN3141** [comment]
> 
> 
> Thanks for jumping in here! This looks great, although I think we can clean up the error handling, and unfortunately the CHANGES note went stale already.

Can see that the changelog has jumped quite a bit since my commit already ahaha. Have cleaned that one up, and will include your suggestions on error handling 👍

File: components/sync15/src/clients/engine.rs (hunk 2 above)
Comment (JN3141) [reply components/sync15/src/clients/engine.rs:77]: Yup, that sounds good; I'll add some `debug` log lines for when we drop them

File: components/sync15/src/clients/engine.rs (hunk 1 above)
Comment (JN3141) [reply components/sync15/src/clients/engine.rs:80]: Ah, yup, using the `match` is quite a bit cleaner for the error handling; will give that a shot.

**Review by mhammond — APPROVED** [review]
This looks great, thanks!

**Pull request merged**